### DISCUSSION TITLE
ergo dash layout

### DISCUSSION THREAD

**Discussion by shirobachi** [nickcoutsos/keymap-editor#138] [Q&A] [answered]
Hey,

Firstly, thanks a lot for the app, it's awesome!
I've ergodash with layout like [that](https://shop.yushakobo.jp/cdn/shop/products/ergodashcase2_1.jpg?v=1650873986) but web show it like ![image](https://github.com/nickcoutsos/keymap-editor/assets/16509749/641255d6-de30-4f34-98d0-4ba9af6c4e75)
Is there a way to let them know how it should be rendered?
Also can I somehow import key map from qmk like [this one](https://config.qmk.fm/#/omkbd/ergodash/rev1/LAYOUT_4key)

Thanks in advance!

**Comment by nickcoutsos** [comment]
Hey there!

So, this tool doesn't have a canonical source for keyboard layout data. It uses the same format that QMK's configurator uses (with some custom extensions) so the same `info.json` used there could be used in the app by saving it in `config/info.json` or `config/ergodash.json` in your repo, _buuut_:

1. The most important thing is that the _order_ of keys in that layout match up with the order of bindings in the ZMK board definition
2. The keymap editor operates by manipulating devicetree code and to do that in a way that's readable it relies on the layout defining `row` and `col` properties to represent the textual layout of the keymap bindings. That requirement is particular to this app so when you're getting a layout from another source it won't be present.

If a layout isn't available the app will try to make one up based on the matrix transform in the ZMK board definition -- that's what we're looking at in the above screenshot. It's meant to be "just enough" to get started, but if customizations are needed or desired, I've been working on some tooling to make it... less painful. [Keymap Layout Tools](https://nickcoutsos.github.io/keymap-layout-tools/) has some code and a helper app to get live feedback when making changes to a layout, and more recently I have added functionality to import more accurate layout data from Kicad PCB files.

Ergodash seems to be a little more complex... there are [three revisions of the PCB](https://github.com/omkbd/ErgoDash/tree/master/PCB) and they all seem to feature variable positions of some keys, but more importantly it seems an entire column of the thumb cluster is optional and the [ZMK definition](https://github.com/zmkfirmware/zmk/blob/main/app/boards/shields/ergodash/ergodash.dtsi) is excluding them.

Do you have any idea how "official" this is? I'd be concerned with importing the more customized version and breaking compatibility with the version supported by ZMK.

**Reply by nickcoutsos** [reply]
In fact, the ZMK version appears to include one of the variable position switches twice. Here's a version imported from a PCB with duplicate keys and the extra thumb columns removed:

![Screen Shot 2023-09-14 at 2 46 06 PM](https://github.com/nickcoutsos/keymap-editor/assets/1165066/24e915e4-3d82-492c-8e76-8295d0bf41c9)

In the matrix transform for the ZMK definition there appears to be 5 keys in the bottom row which I don't think corresponds to any singular switch placement. Keymap Editor aside, can you confirm that the ZMK layout works as expected? Someone might need to supply alternate matrix transforms for other setups but I could see that getting complicated.

**Reply by shirobachi** [reply]
Hey, 

So at first thanks for so much information, I am totally newbie, actually. I bought this made up keyboard for me and just starting using it, or making it work, but I have programming background, I am not that crazy 😅
> "Do you have any idea how "official" this is?"

So about that I don't really know. However, qmk have only two versions and one of them is `mini` and other is `rev1` (same as in ergo dash repo what you linked) here is [repo](https://github.com/qmk/qmk_firmware/tree/master/keyboards/omkbd/ergodash). Does it somehow help?
Also, maybe this picture will help. If there is no arrow is as expected and tested only master (left) by now, luck of time ;/
![image](https://github.com/nickcoutsos/keymap-editor/assets/16509749/7a8909eb-be43-475a-9724-27be979a578f)
Just for reference, mine looks like [that](https://github.com/nickcoutsos/keymap-editor/assets/16509749/6ee665f5-7c48-456a-96a4-edfeff1416d8) one.

> [Keymap Layout Tools](https://nickcoutsos.github.io/keymap-layout-tools/)

Thanks for that link, I'll play around with that tool a bit <3 

> (...) can you confirm that the ZMK layout works as expected?

So at least left master and with dump test. On the screenshot, you may observe that there is no "change mode" key, not sure if this make any changes, but if it does now you know.


Thanks a lot of for your help🤩
Simon

**Reply by nickcoutsos** [reply, marked as answer]
I wouldn't worry too much about a mapping from the QMK Configurator diagram as we aren't looking at a default keymap implemented in two firmwares (also the QMK diagram includes the extra thumb columns).

I'm not sure what the "change mode" key is but I'm noticing that one of the variable position keys is `&none`. Now I'm suspecting that each of the two switch positions exist separately in the matrix. That's a little clumsy but it's a problem for representing visually in the Keymap Editor as you'd need to make those keys invisible or else just ignore them.

**Reply by shirobachi** [reply]
Sorry “change mode” I meant change layer, my bad. Oh, yeah, there is `&none` I just did not change it yet.
From my perspective we can drop this, I was wondering if maybe that just easy, but that's no problem since I know how to check what is what.


Thanks a lot, anyway, 
If I could somehow help, feel free to ping me!
Simon

**Reply by shirobachi** [reply]
Could I ask for a little help?
If so could you please have a look at [this file](https://github.com/shirobachi/zmk-config-nrfmicro/blob/a2c0bc12243757365b08cefc480d9dc40b08b007/config/info.json)? Is it correct "info.json" file by syntax? I got [some errors](https://termbin.com/7kdd) however your [Keymap Layout Helper](https://nickcoutsos.github.io/keymap-layout-tools/) does not complain at all


Thanks in advance,
Simon

**Reply by nickcoutsos** [reply]
I think I may have unintentionally added some validation checks to the editor that I hadn't yet finished implementing in the layout tools helper. As you can see, it's a lot of errors that all kind of mean the same thing but aren't that easy to take in.

Pasting your layout into the helper app I get the following textual rendering:

```
 0   1   2   3   4   5   6
14  15  16  17  18  19  20
28  29  30  31  32  33  34
42  43  44  45  46  47  48
56  57  58  59  60  61  62
    12  11  10   9   8   7  13
    26  25  24  23  22  21  27
    40  39  38  37  36  35  41
    54  53  52  51  50  49  55
    68  67  66  65  64  63  69
```

This should roughly match the visual layout but it appears that the left and right keyboard halves are separated. Additionally, it looks like some of the columns (but not all) are reversed in the right half.

To fix this I use the `Re-order` action to visually map keys that are in the same approximate row or column.

<img width="887" alt="Screen Shot: Re-order tool showing the current ergodash row/col mapping" src="https://github.com/nickcoutsos/keymap-editor/assets/1165066/df4d5941-51af-4711-88a7-9cfc75fcaeca">

From this point it's simplest to start over by hitting the _Clear_ button and creating now row/column groups.

  
<img width="877" alt="Screen Shot 2023-10-08 at 12 20 55 PM" src="https://github.com/nickcoutsos/keymap-editor/assets/1165066/ed0d2a96-1d92-43f8-95a3-d85f5fd193e1">

This version is a little better organized, but some guesses had to be made with the thumb clusters:

```
 0   1   2   3   4   5   6                     7   8   9  10  11  12  13
14  15  16  17  18  19  20                    21  22  23  24  25  26  27
28  29  30  31  32  33  34                    35  36  37  38  39  40  41
42  43  44  45  46  47      48            49      50  51  52  53  54  55
56  57  58  59          60  61  62    63  64  65          66  67  68  69
```

<img width="830" alt="Screen Shot 2023-10-08 at 12 25 04 PM" src="https://github.com/nickcoutsos/keymap-editor/assets/1165066/ad728808-97fc-4933-b13d-1efc2cab48dd">

The other consideration is that this ultimately needs to match the matrix transform for this keyboard

https://github.com/zmkfirmware/zmk/blob/fd05478897567bff34d607229636ae4d4507ef0f/app/boards/shields/ergodash/ergodash.dtsi#L27-L34

```
        map = <
RC(0,0) RC(0,1) RC(0,2) RC(0,3) RC(0,4) RC(0,5) RC(0,6)                 RC(0,13) RC(0,12) RC(0,11) RC(0,10) RC(0,9) RC(0,8) RC(0,7)
RC(1,0) RC(1,1) RC(1,2) RC(1,3) RC(1,4) RC(1,5) RC(1,6)                 RC(1,13) RC(1,12) RC(1,11) RC(1,10) RC(1,9) RC(1,8) RC(1,7)
RC(2,0) RC(2,1) RC(2,2) RC(2,3) RC(2,4) RC(2,5) RC(2,6)                 RC(2,13) RC(2,12) RC(2,11) RC(2,10) RC(2,9) RC(2,8) RC(2,7)
RC(3,0) RC(3,1) RC(3,2) RC(3,3) RC(3,4) RC(3,5)         RC(3,6) RC(3,13)         RC(3,12) RC(3,11) RC(3,10) RC(3,9) RC(3,8) RC(3,7)
RC(4,0) RC(4,1) RC(4,2) RC(4,3) RC(4,4)         RC(4,5) RC(4,6) RC(4,13) RC(4,12)         RC(4,11) RC(4,10) RC(4,9) RC(4,8) RC(4,7)
        >;
  };
```

There are two discrepancies:

1. Your layout is omitting that extra switch in the bottom row for the alternate switch placement
2. Your layout includes an extra switch for the 2u thumb keys

For the first issue I think you'd just have to include a key for both key placements and just ignore one.

For the second, are there any `RC(...)` entries in the matrix transform that map to the switches where your 2u keys would go?

**Reply by shirobachi** [reply]
Oh god, it's big, 
So I managed to copy what you do (I like helper, it's very easy to use :) ) Do you consider adding option for manual type `x`, `y` position? I would vote for that feature:)

> Your layout is omitting that extra switch in the bottom row for the alternate switch placement
Unfortunately, I am not sure what you mean? Is it about my layout's last row has 4 keys when on the zmk is 5? I believe it's on of thump key is on this row, see [here](https://github.com/nickcoutsos/keymap-editor/assets/16509749/c9e6c7a6-d8ed-4b0a-979c-b2c1897eb207)

Also, is there any way to make sure that info.json is recognized? I am suspecting that this is not always taken (like I got error before this what I sent, but Today I don't get an error when I'm expecting this)

> Your layout includes an extra switch for the 2u thumb keys
I thought it won't be a problem, but I think it should be (4,6) for left one or (4,13) for right one


Also, sorry for wasting your time, I thought that since visually it was right it would be simple missing comma or sth similar, not that big. Since I am newbie, I believe that Ill not manage that ;/ 



Thanks for your help a lot! 
Simon

**Comment by nickcoutsos** [comment]
> Do you consider adding option for manual type `x`, `y` position? I would vote for that feature:)

This is possible already, the panel on the left is the raw layout data and you can edit `x`/`y`/`row`/`col` etc. Did you have something else in mind?

> Also, is there any way to make sure that info.json is recognized? I am suspecting that this is not always taken (like I got error before this what I sent, but Today I don't get an error when I'm expecting this)

I see you have `ergodash.json` as well as `info.json` in your repo. The more specific `<keyboard>.json` is preferred to support repositories where multiple keymaps are being edited, and `info.json` is used as the fallback. So if you'd prefer the latter either delete `ergodash.json` or replace it with `info.json`.

**Reply by shirobachi** [reply]
Oh I meant, in reorder UI, but actually I forgot that you there is this code block, work for me – thanks

Oh, so I checked now, and it works, just keys are messed up, I'll fix it – thanks!

**Reply by nickcoutsos** [reply]
I can't remember if I mentioned it before but with the differences between what you want and what the matrix transform allows I would recommend not worrying too much about the app right now. Focus on making sure your keyboard is set up correctly and that every key you bind something to actually triggers the expected behavior, then we can make adjust a layout to match it.

**Reply by shirobachi** [reply]
I got it, it works as supposed even visually, but what fight it was 😆

One more time, thanks a loooot for you time and help, hope I did not frustrate you too much 😅

**Reply by nickcoutsos** [reply]
No worries, I get that this is a lot to take in when you want things to just work. It's unfortunate that the default configuration couldn't have made things simpler but maybe it comes down to complexity in the PCB layout.

If everything is working as expected then I have a tweaked version of the layout to hopefully make the bindings array a little easier to read:

```json
{
  "id": "ergodash",
  "name": "ergodash",
  "layouts": {
    "LAYOUT": {
      "layout": [
        { "row": 0, "col":  0, "x":    0, "y": 0.375 },
        { "row": 0, "col":  1, "x":    1, "y": 0.375 },
        { "row": 0, "col":  2, "x":    2, "y": 0.125 },
        { "row": 0, "col":  3, "x":    3, "y":     0 },
        { "row": 0, "col":  4, "x":    4, "y": 0.125 },
        { "row": 0, "col":  5, "x":    5, "y":  0.25 },
        { "row": 0, "col":  6, "x":    6, "y":  0.75 },
        { "row": 0, "col": 12, "x": 11.5, "y":  0.75 },
        { "row": 0, "col": 13, "x": 12.5, "y":  0.25 },
        { "row": 0, "col": 14, "x": 13.5, "y": 0.125 },
        { "row": 0, "col": 15, "x": 14.5, "y":     0 },
        { "row": 0, "col": 16, "x": 15.5, "y": 0.125 },
        { "row": 0, "col": 17, "x": 16.5, "y": 0.375 },
        { "row": 0, "col": 18, "x": 17.5, "y": 0.375 },
      
        { "row": 1, "col":  0, "x":    0, "y": 1.375 },
        { "row": 1, "col":  1, "x":    1, "y": 1.375 },
        { "row": 1, "col":  2, "x":    2, "y": 1.125 },
        { "row": 1, "col":  3, "x":    3, "y":     1 },
        { "row": 1, "col":  4, "x":    4, "y": 1.125 },
        { "row": 1, "col":  5, "x":    5, "y":  1.25 },
        { "row": 1, "col":  6, "x":    6, "y":  1.75 },
        { "row": 1, "col": 12, "x": 11.5, "y":  1.75 },
        { "row": 1, "col": 13, "x": 12.5, "y":  1.25 },
        { "row": 1, "col": 14, "x": 13.5, "y": 1.125 },
        { "row": 1, "col": 15, "x": 14.5, "y":     1 },
        { "row": 1, "col": 16, "x": 15.5, "y": 1.125 },
        { "row": 1, "col": 17, "x": 16.5, "y": 1.375 },
        { "row": 1, "col": 18, "x": 17.5, "y": 1.375 },
      
        { "row": 2, "col":  0, "x":    0, "y": 2.375 },
        { "row": 2, "col":  1, "x":    1, "y": 2.375 },
        { "row": 2, "col":  2, "x":    2, "y": 2.125 },
        { "row": 2, "col":  3, "x":    3, "y":     2 },
        { "row": 2, "col":  4, "x":    4, "y": 2.125 },
        { "row": 2, "col":  5, "x":    5, "y":  2.25 },
        { "row": 2, "col":  6, "x":    6, "y":  2.75 },
        { "row": 2, "col": 12, "x": 11.5, "y":  2.75 },
        { "row": 2, "col": 13, "x": 12.5, "y":  2.25 },
        { "row": 2, "col": 14, "x": 13.5, "y": 2.125 },
        { "row": 2, "col": 15, "x": 14.5, "y":     2 },
        { "row": 2, "col": 16, "x": 15.5, "y": 2.125 },
        { "row": 2, "col": 17, "x": 16.5, "y": 2.375 },
        { "row": 2, "col": 18, "x": 17.5, "y": 2.375 },
      
        { "row": 3, "col":  0, "x":    0, "y": 3.375 },
        { "row": 3, "col":  1, "x":    1, "y": 3.375 },
        { "row": 3, "col":  2, "x":    2, "y": 3.125 },
        { "row": 3, "col":  3, "x":    3, "y":     3 },
        { "row": 3, "col":  4, "x":    4, "y": 3.125 },
        { "row": 3, "col":  5, "x":    5, "y":  3.25 },
        { "row": 3, "col":  7, "x":    6, "y":     4 },
        { "row": 3, "col": 11, "x": 11.5, "y":     4 },
        { "row": 3, "col": 13, "x": 12.5, "y":  3.25 },
        { "row": 3, "col": 14, "x": 13.5, "y": 3.125 },
        { "row": 3, "col": 15, "x": 14.5, "y":     3 },
        { "row": 3, "col": 16, "x": 15.5, "y": 3.125 },
        { "row": 3, "col": 17, "x": 16.5, "y": 3.375 },
        { "row": 3, "col": 18, "x": 17.5, "y": 3.375 },
      
        { "row": 4, "col":  0, "x":    0, "y": 4.375 },
        { "row": 4, "col":  1, "x":    1, "y": 4.375 },
        { "row": 4, "col":  2, "x":    2, "y": 4.125 },
        { "row": 4, "col":  3, "x":    3, "y":     4 },
        { "row": 4, "col":  6, "x":    5, "y":     5 },
        { "row": 4, "col":  7, "x":    6, "y":     5 },
        { "row": 4, "col":  8, "x":    7, "y":     4, "h": 2 },
        { "row": 4, "col": 10, "x": 10.5, "y":     4, "h": 2 },
        { "row": 4, "col": 11, "x": 11.5, "y":     5 },
        { "row": 4, "col": 12, "x": 12.5, "y":     5 },
        { "row": 4, "col": 15, "x": 14.5, "y":     4 },
        { "row": 4, "col": 16, "x": 15.5, "y": 4.125 },
        { "row": 4, "col": 17, "x": 16.5, "y": 4.375 },
        { "row": 4, "col": 18, "x": 17.5, "y": 4.375 }
      ]
    }
  },
  "sensors": []
}
```

Results in:

```
 0   1   2   3   4   5   6                     7   8   9  10  11  12  13
14  15  16  17  18  19  20                    21  22  23  24  25  26  27
28  29  30  31  32  33  34                    35  36  37  38  39  40  41
42  43  44  45  46  47      48            49      50  51  52  53  54  55
56  57  58  59          60  61  62    63  64  65          66  67  68  69
```

**Reply by Doc-Worm** [reply]
Hello everyone.  Complete newbie here with no programing background.  I'm trying to get my ergodash running with ZMK since i'm using wireless controllers - But as has been noted, the thumb clusters are wonky.  For he thumb cluster, my configuration uses 3 keys on the left and 2 on the right (see attached).

Can anyone help me out by giving me some direction on how to create this layout?  Thanks, I really appreciate any advice!
![20231021_175939](https://github.com/nickcoutsos/keymap-editor/assets/146211165/cf8f8a7f-7a3e-478e-a59d-58dba20ecbb7)

**Reply by nickcoutsos** [reply]
Hi @Doc-Worm can you link to a repository where this is set up? The default in ZMK isn't set up that way so some changes may be required in your version.

**Reply by Doc-Worm** [reply]
Hello @nickcoutsos.  i think this is what you're requesting:  https://github.com/Doc-Worm/zmk-config

is that right?

**Reply by nickcoutsos** [reply]
Hi, yeah, that's it. So, without a custom `config/ergodash.dtsi` in your repo you'll be building from [the default version](https://github.com/zmkfirmware/zmk/blob/main/app/boards/shields/ergodash/ergodash.dtsi) which uses 3 1u thumb keys in each hand. How well do you know the schematic for this board and its various switch layouts? You'll need to create a matrix transform with the correct `RC(...)` values, but I don't know what they're supposed to map to.

**Reply by Doc-Worm** [reply]
I'm not entirely sure i know what a schematic for this board is . . . does that answer how well i know it?  I know that [these](https://ergotaiwan.tw/wp-content/uploads/ergodash-layout-number-1024x901.jpg) are the possible configurations.  Not sure what a matrix transform is but I gather we want to change the numbers after the RC for the thumb cluster to whatever they're supposed to be?

> 
RC(0,0) RC(0,1) RC(0,2) RC(0,3) RC(0,4) RC(0,5) RC(0,6)                 RC(0,13) RC(0,12) RC(0,11) RC(0,10) RC(0,9) RC(0,8) RC(0,7)
RC(1,0) RC(1,1) RC(1,2) RC(1,3) RC(1,4) RC(1,5) RC(1,6)                 RC(1,13) RC(1,12) RC(1,11) RC(1,10) RC(1,9) RC(1,8) RC(1,7)
RC(2,0) RC(2,1) RC(2,2) RC(2,3) RC(2,4) RC(2,5) RC(2,6)                 RC(2,13) RC(2,12) RC(2,11) RC(2,10) RC(2,9) RC(2,8) RC(2,7)
RC(3,0) RC(3,1) RC(3,2) RC(3,3) RC(3,4) RC(3,5)         RC(3,6) RC(3,13)         RC(3,12) RC(3,11) RC(3,10) RC(3,9) RC(3,8) RC(3,7)
RC(4,0) RC(4,1) RC(4,2) RC(4,3) RC(4,4)         RC(4,5) RC(4,6) RC(4,13) RC(4,12)         RC(4,11) RC(4,10) RC(4,9) RC(4,8) RC(4,7)


I can tell you that currently, on the left side, the only button that does anything for the thumb cluster is the "page down" button (see pic above).  It seems to be mapped to spacebar.
on the right side, the black key is mapped to enter but the enter button doesn't do anything.

Hope that helps.  Sorry, I'm trying to learn this stuff but have no background with anything like this, so it is a bit of a steep curve.

thanks!

**Reply by nickcoutsos** [reply]
Having the identifiers for each combination of key configurations is helpful as a concept, but to your point we'd need to know what row and column each switch would map to. Do you have a vendor or manufacturer that could support this? I think the original ZMK ergodash definition was provided by a random user and tailored to their particular layout.

**Reply by Doc-Worm** [reply]
I just ordered a parts kit for the board and then sourced controllers, batteries, switches, and caps, so no manufacturer or anything.  I put it all together (another first for me, but fun to learn).  The board is the ergodash rev. 1.2.

**Reply by nickcoutsos** [reply]
\*sigh\* it's pretty tough.

I've spent a bunch of time last night trying to correlate the ZMK definition, the QMK definition, and imported PCB data but some things aren't quite adding up. I'm not 100% convinced that the existing ZMK definition is correct, even accounting for the possibility of different switch locations.

Looking at all of the Kicad data might help resolve this but that's beyond my knowledge and tools. I'm quite preoccupied for the next couple of days but after that I might come back to this and try something with a little less dependent on existing implementations. Some of it might be guesswork, and I have nothing to test it on, are you ok with trying a bunch of random crap? Do you happen to have a multi meter you could use to confirm which switches connect to which pins?

**Reply by Doc-Worm** [reply]
I appreciate your help.


In an ideal world I could get it configured like I have it, but I didn't realize it would be such a pain in the butt.  I could just resolder the thumb clusters to match the default layout .

I do have a multimeter, but would need instructions on how to test it.  I'm all about trying random crap, but want to be conscious of your time and not take advantage of your generosity in helping.

edit:
also, i should note that i just tried a little app that displays what key you're pressing and it turns out the only key that don't work on the thumb clusters is the right innermost key (where i put the enter key)

**Reply by Doc-Worm** [reply]
So I re-soldered it to match the layout and that worked!  Thanks for your help though, it has been a learning experience.  Now I just need to get used to typing on an ortholinear board.  Also, your tool for mapping is really good, thanks.

**Reply by nickcoutsos** [reply]
That's great! What does your layout look like now?

**Reply by Doc-Worm** [reply]
here it is:
![20231024_190542](https://github.com/nickcoutsos/keymap-editor/assets/146211165/05d7c8c6-eec5-4bec-bd02-b3daf99cc5e6)

**Reply by nickcoutsos** [reply]
Ah, right, matching the layout for the board defined in ZMK. So you've still got the extra bindings in the bottom row where there's two possible positions (_Super_ and _End_ in your photo), but I guess right now what matters most is that the keyboard works.

**Reply by Doc-Worm** [reply]
not sure what you mean by extra bindings, as the physical keyboard and the map matched up 1:1 . . but you're right, it working is all that i really care about.  If i do this again i might have to get deeper in the weeds of this but for now i'm good.  

I've been using your tool though to change the mappings as I learn what keys i'm missing that i use and what the optimal place for the space bar is, etc.  Now i just need to figure out how to replace the text on some of the caps with what they are actually mapped to.  I'm thinking stickers.

**Reply by nickcoutsos** [reply]
Sorry, I might have made a mistake. It does seem that all of the keys are in use, although the graphical representation doesn't use the 2h keys and I think they're ordered differently. Is that right?

**Comment by Doc-Worm** [comment]
the 2u key is the bottom right key, the other two are the other two.  now i just need to figure out why battery life is so bad.  the battery life estimator for zmk says i should get a few months out of the right side when in reality I only get a few days.
